### PR TITLE
[Feat][Router] add model aliases

### PR DESCRIPTION
**Pull request by max-wittig**
This feature allows:
* to preserve compatibility of model names when newer models are released
* allow users to specify known commercial models and route the requests to open weight models instead

E.g. if I want to replace `mistral-7b-instruct` with the more powerful `mistral-nemo-instruct-2407` and not disrupt the users, and at the same time route `gpt-4` to `mistral-nemo-instruct-2407` I could just specify it like this

```sh
python3 -m vllm_router.app \
--static-aliases "mistral-7b-instruct:mistral-nemo-instruct-2407,gpt-4:mistral-nemo-instruct-2407"
```

FIX https://github.com/vllm-project/production-stack/issues/389

---

- [x] Make sure the code changes pass the [pre-commit](../CONTRIBUTING.md) checks.
- [x] Sign-off your commit by using <code>-s</code> when doing <code>git commit</code>
- [x] Try to classify PRs for easy understanding of the type of changes, such as `[Bugfix]`, `[Feat]`, and `[CI]`.

<details>

<summary><b> Detailed Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to production-stack! Before submitting the pull request, please ensure the PR meets the following criteria. This helps us maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Please try to classify PRs for easy understanding of the type of changes. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Feat]</code> for new features in the cluster (e.g., autoscaling, disaggregated prefill, etc.).</li>
    <li><code>[Router]</code> for changes to the <code>vllm_router</code> (e.g., routing algorithm, router observability, etc.).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>Pass all linter checks. Please use <code>pre-commit</code> to format your code. See <code>README.md</code> for installation.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li> Please include sufficient tests to ensure the change is stay correct and robust. This includes both unit tests and integration tests.</li>
</ul>

<h3>DCO and Signed-off-by</h3>
<p>When contributing changes to this project, you must agree to the <a href="https://github.com/vllm-project/vllm/blob/main/DCO">DCO</a>. Commits must include a <code>Signed-off-by:</code> header which certifies agreement with the terms of the DCO.</p>
<p>Using <code>-s</code> with <code>git commit</code> will automatically add this header.</p>

<h3>What to Expect for the Reviews</h3>

We aim to address all PRs in a timely manner. If no one reviews your PR within 5 days, please @-mention one of YuhanLiu11
, Shaoting-Feng or ApostaC.

### CODE REVIEW

File: src/vllm_router/services/request_service/request.py [hunk 1]
```diff
@@ -194,12 +195,19 @@ async def route_general_request(
...
     engine_stats = request.app.state.engine_stats_scraper.get_engine_stats()
     request_stats = request.app.state.request_stats_monitor.get_request_stats(
         time.time()
     )
 
+    aliases = getattr(service_discovery, "aliases", None)
+    if aliases and requested_model in aliases.keys():
+        requested_model = aliases[requested_model]
+        request_body = replace_model_in_request_body(request, request_body, request_json, requested_model)
+        update_content_length(request, request_body)
```
Comment (max-wittig) [inline src/vllm_router/services/request_service/request.py:210]: I doubt that the `request_rewriter` above works without this part.

**Comment by max-wittig** [comment]
Before I add documentation, tests etc. Could you just give it a brief review just that it goes in the right direction?

@YuhanLiu11

**Comment by YuhanLiu11** [comment]
@max-wittig Thanks for the PR! A higher level question, could you elaborate on why using vLLM’s ``--served-model-name`` doesn’t work for your use case?

**Comment by max-wittig** [comment]
@YuhanLiu11 Our deployment runs on multiple machines with different kind of configurations. The GPU instances are real physical machines in a data center, some on AWS etc, which are harder to adjust and update than just adjusting the vllm-router. It allows us to have the configuration deployed in one place and be effective.

Basically its nice for faster iterations and it seems more natural to have this part inside the router.

**Comment by YuhanLiu11** [comment]
@max-wittig I see. In that case this PR looks good on a high level (if it can pass the tests).

**Comment by max-wittig** [comment]
@YuhanLiu11 Thanks for the comments! I will fix the tests and add some for the functionality then!

**Comment by max-wittig** [comment]
I didn't add tests for the service router as currently there are none yet, but I'm open to some in general in a follow-up.

**Comment by YuhanLiu11** [comment]
@max-wittig thanks! I'll take a closer look at this PR later today or tomorrow (since this is a relatively large one)!

File: docs/requirements-docs.txt [hunk 2]
```diff
@@ -6,3 +6,4 @@ sphinx-copybutton==0.5.2
 sphinx-togglebutton==0.3.2
 sphinx_design==0.6.1
 sphinxemoji==0.3.1
+starlette==0.45.3
```
Comment (pwuersch) [inline docs/requirements-docs.txt:9]: @max-wittig is adding this package here required? Firstly, it's a change in the documentation requirements, and secondly, `starlette` is already included when installing `fastapi`, see https://github.com/fastapi/fastapi/blob/9a33ba46ac3a3ef0b845c8515822f4e13c0fb443/pyproject.toml#L46. I believe `fastapi` is already installed in all actions where the tests are run. Or was this due to some other change / failing CI step?

Comment (max-wittig) [reply docs/requirements-docs.txt:9]: yes. Otherwise the pipeline breaks.

File: src/vllm_router/services/request_service/request.py [hunk 3]
```diff
@@ -195,12 +196,19 @@ async def route_general_request(
...
     engine_stats = request.app.state.engine_stats_scraper.get_engine_stats()
     request_stats = request.app.state.request_stats_monitor.get_request_stats(
         time.time()
     )
 
+    aliases = getattr(service_discovery, "aliases", None)
+    if aliases and requested_model in aliases.keys():
+        requested_model = aliases[requested_model]
+        request_body = replace_model_in_request_body(request_json, requested_model)
+        update_content_length(request, request_body)
```
Comment (pwuersch) [inline src/vllm_router/services/request_service/request.py:210]: This double-lookup into `aliases` could be avoided. WDYT? It is a hot-path that is executed on every request, so this might be worth optimizing although the lookup is not that expensive.

```suggestion
    if aliases:
        if actual_model := aliases.get(requested_model):
            request_body = replace_model_in_request_body(request_json, actual_model)
            update_content_length(request, request_body)
```

**Review by pwuersch — CHANGES_REQUESTED** [review]
@max-wittig I went over your changes and found one small improvement that could be done. Other than that, the feature looks good to me and solves our use-case as expected. Thank you very much for implementing it.

File: src/vllm_router/services/request_service/request.py (hunk 3 above)
Comment (max-wittig) [reply src/vllm_router/services/request_service/request.py:210]: in my opinion this makes it a bit harder to read, but I will leave it up to the maintainers.

Comment (pwuersch) [reply src/vllm_router/services/request_service/request.py:210]: agree 👍🏻

File: src/vllm_router/services/request_service/request.py [hunk 4]
```diff
@@ -195,12 +196,19 @@ async def route_general_request(
             logger.warning("Failed to parse rewritten request body as JSON")
 
     # TODO (ApostaC): merge two awaits into one
-    endpoints = get_service_discovery().get_endpoint_info()
+    service_discovery = get_service_discovery()
+    endpoints = service_discovery.get_endpoint_info()
```
Comment (YuhanLiu11) [inline src/vllm_router/services/request_service/request.py:200]: Why do we need to split this line into two?

File: src/vllm_router/services/request_service/request.py (hunk 3 above)
Comment (YuhanLiu11) [reply src/vllm_router/services/request_service/request.py:210]: The current code looks good to me

File: src/vllm_router/services/request_service/request.py (hunk 4 above)
Comment (max-wittig) [reply src/vllm_router/services/request_service/request.py:200]: Because I need to use the service discovery variable later in line 206 to get the aliases, if they exist:

https://github.com/vllm-project/production-stack/pull/397/files#diff-3e852cfcf107dfd78ee2b7c7e88e05428b7f2bffcfc5ca1dce0c070caf563f2aR206

Let me know, if there is a better way to do things.

Comment (YuhanLiu11) [reply src/vllm_router/services/request_service/request.py:200]: I see. ok. This looks good then.

**Review by YuhanLiu11 — APPROVED** [review]
LGTM, thanks for the feature!
